sympy/matrices/common.py

```diff
@@ -2218,37 +2218,42 @@ def __neg__(self):
         return self._eval_scalar_mul(-1)
 
     @call_highest_priority('__rpow__')
-    def __pow__(self, num):
+    def __pow__(self, exp):
         if self.rows != self.cols:
             raise NonSquareMatrixError()
         a = self
-        jordan_pow = getattr(a, '_matrix_pow_by_jordan_blocks', None)
-        num = sympify(num)
-        if num.is_Number and num % 1 == 0:
+        exp = sympify(exp)
+        jordan_pow = getattr (a, '_matrix_pow_by_jordan_blocks', None)
+
+        if exp.is_Number and exp % 1 == 0:
             if a.rows == 1:
-                return a._new([[a[0]**num]])
-            if num == 0:
+                return a._new([[a[0]**exp]])
+            if exp == 0:
                 return self._new(self.rows, self.cols, lambda i, j: int(i == j))
-            if num < 0:
-                num = -num
+            if exp == 1:
+                return self
+            if exp < 0:
+                exp = -exp
                 a = a.inv()
             # When certain conditions are met,
             # Jordan block algorithm is faster than
             # computation by recursion.
-            elif a.rows == 2 and num > 100000 and jordan_pow is not None:
+            elif a.rows == 2 and exp > 100000 and jordan_pow:
                 try:
-                    return jordan_pow(num)
+                    return jordan_pow(exp)
                 except MatrixError:
-                    pass
-            return a._eval_pow_by_recursion(num)
-        elif not num.is_Number and num.is_negative is None and a.det() == 0:
-            from sympy.matrices.expressions import MatPow
-            return MatPow(a, num)
-        elif isinstance(num, (Expr, float)):
-            return jordan_pow(num)
-        else:
-            raise TypeError(
-                "Only SymPy expressions or integers are supported as exponent for matrices")
+                    pass # a may not have _matrix_pow_by_jordan_blocks?
+            return a._eval_pow_by_recursion(exp)
+
+        if jordan_pow:
+            try:
+                return jordan_pow(exp)
+            except ValueError:
+                if exp.is_negative or exp.is_number:
+                    raise
```

Comment:
Why catch the error and then raise in these particular cases? What are the other cases where it is good to catch and ignore the error?

Reply:
Because otherwise this would raise:
```
A = Matrix([[0, 1, 0], [0, 0, 1], [0, 0, 1]])  # Nilpotent jordan block size 2
n = Symbol('n', integer=True)
assert isinstance(A**n, MatPow)
```

Reply:
Where is the ValueError raise from?

Reply:
_matrix_pow_by_jordan_blocks() detects the matrix is determinant zero and bails but in this case the symbol is not defined positive or negative so we defer raising until it is actually substituted with something negative, otherwise substituting positive integer should be fine

Reply:
Not necessarily a change for this PR but it would be better to raise (and catch) something more specific like NotInvertibleMatrixError for this.

Reply:
That is a change I wouldn't make unless necessary since it is not really functional but more aesthetic but has the potential to break people's code.

Reply:
It's fine if you don't want to change that in this PR.

Note that it can be changed in a backwards compatible way if the new exception class is a subclass of the old. Also it's not purely aesthetic: overly broad exception handling genuinely hides bugs. We shouldn't use generic exceptions like ValueError for exceptional cases that are expected to occur by design.

Reply:
I was thinking more if people were using the text of the exception to differentiate from the various other uses of MatrixError but I suppose if I keep the text the same there would be now problem. Though looking at it there are only 5 places in the whole project where MatrixError is raised and 3 of them are because of defective matrix and the other 2 for internal errors. If you really want to differentiate that I guess its not a big deal...

Reply:
Users should not depend on the text of exceptions. That should be obviously fragile and we really can't support that in a backwards compatible way.

MatrixError is the superclass for the more specific errors ShapeError, NonSquareMatrixError, and NonPositiveDefiniteMatrixError which are raised in many places. These also subclass ValueError so can be caught by that as well. I would suggest a new error: NonInvertibleMatrixError.

Mostly there isn't a need to catch these exceptions but if (as here) we do want to do that as part of normal control flow then it really needs to be something much more precise than ValueError.

Reply:
They shouldn't, but they do, in the end its all about how much code breaks, but you are right in principle here and I can do it in a non-breaking way so will do.